New logger: JSON diff output

thomasjm opened the pull request:
# Motivation

This introduces a new logger which outputs the build state in the form of RFC 6902 JSON diffs. The first line is a JSON object like this: `{"activities":{},"messages":[]}`, and subsequent lines are diffs which transform this object. These diffs are computed every 300ms, which works well but could be made configurable.

The motivation for this is to output Nix build information in a machine-readable way for downstream tooling to consume, so the build state can be shown in a nice UI. Why not use the existing `internal-json` log format? Firstly because it is a little bit wonky/internal in its output, and secondly because it produces quite a bit of output. The amount of compute time for a downstream tool to parse all the JSON and apply it to its in-memory representation can be very expensive. The advantage of the current PR is that a series of quick updates cause no additional encoding/decoding expense, as the Nix tool's in-memory build state gets updated. Then, that data structure is periodically diffed and the changes printed. In practice this can be a large win.

I was hoping the Nix team might be willing to merge this to support the kind of tooling I have in mind. Here's a screenshot of this in action in the tool I'm working on:

![environment_install_demo](https://user-images.githubusercontent.com/1634990/216883978-01c2d1e0-9d66-4dbb-b89c-51439abfdbc8.gif)



# Context

Non-trivial change: this logger was meaty enough that I put it in its own files, `logging-diffs.cc/hh`. The other loggers are in `logging.cc/hh`.

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or bug fix: updated release notes

cole-h reviewed (commented):
Interesting idea. Maybe not entirely similar, but this also reminded me of https://github.com/lf-/nix-otel.

(Disclaimer: I'm not part of the Nix team, just a bystander with some code ideas / questions / comments.)

thomasjm commented on src/libutil/logging-diffs.hh:
```diff
@@ -0,0 +1,51 @@
+
+namespace nix {
+
+Logger * makeJSONLogger(Logger & prevLogger);
```
Thanks, fixed!

thomasjm commented on src/libutil/logging-diffs.cc:
```diff
@@ -0,0 +1,228 @@
+#include <nlohmann/json.hpp>
+#include <thread>
+
+using json = nlohmann::json;
```
Okay, changed this file to use this alias everywhere for fewer occurrences of `nlohmann::`.

thomasjm commented on src/libutil/logging-diffs.cc:
```diff
@@ -0,0 +1,228 @@
+
+            sendLatestIfNecessary();
+
+            std::this_thread::sleep_for(std::chrono::milliseconds(300));
```
I'd suggest `diff-logger-period-ms` or `diff-logger-interval-ms`, since the number is a period and since I like to include units on time intervals :)

thomasjm commented:
Thanks @cole-h! Just responded to your review.

thufschmitt commented:
That is quite interesting :)

Correct me if I'm wrong, but it looks like this is adding two very useful things, that are mixed in the implementation but logically orthogonal:

1. A JSON logger meant for external consumption (compared to the `internal-json` one which was exposed but not really nice to use and with zero stability guaranty)
2. A diff version of a JSON logger to improve the performances.

I really like both, but maybe it could be good to clearly separate them for a couple of reasons:
1. The json-diff output is something that could also benefit the current logger, so it would be great to have it optionally applied here too,
2. If we have a new better JSON output, it would probably be useful to have both a json-diff and a “plain” version of it since I assume that different clients might or might not want the diff version,
3. From a maintainer perspective, the json-diff part is something that we might just merge as the design space here is mostly trivial, so there's not much to debate on. On the other hand, a whole new JSON logger would likely require more time and iterations to be really considered stable.

/cc @maralorn since he's the only current consumer of the `internal-json` format that I can think of.

maralorn commented:
Thanks for the ping. The crucial consideration from the nom perspective here would be whether this output contains all the information. e.g. do messages appear in the diff? Also, 300ms might be a bit sluggish.

Apart from that, this might be greatly beneficial to noms performance. Currently, the biggest bottleneck is performantly parsing the ton of output that nix gives us (primarily the very frequent updates about download state).

On the other hand, the output is still missing information. e.g. currently it is not possible to tell from the output which files, if any, are currently being evaluated (there is only a message when a file was evaluated). Also, currently, it’s not possible to tell from the logs which nar-file download belongs to which store-path. That info is only available from the narinfo file, which nix downloads and keeps in memory. Just putting that out there while we are on it. It does not strictly have anything to do with this PR.

thomasjm commented:
> do messages appear in the diff?

Yep!

> Also, 300ms might be a bit sluggish.

There's a review comment above about this -- we'd likely make it configurable if we move forward with this.

> it looks like this is adding two very useful things, that are mixed in the implementation but logically orthogonal

I'm not sure I quite agree with this framing. I'd say the existing `internal-json` logger is already "diff-like" in the sense that it consists of event data--if you know what all the events mean, you can build up a representation of the build state and update it incrementally on each event. What the `internal-json` logger lacks is consolidation/throttling of the event messages, so it gives you a firehose of events which pretty much represent raw internal events.

You could imagine running the diff logger in the limit where the period goes to zero, so it logs every time an event happens. That would produce a logger similar to `internal-json`, but it wouldn't exactly be event data anymore. I'd say that diff-based logging and event-based logging both have their uses. The former is nice when a downstream tool wants to maintain/display an overall view of the build state, while the latter is nicer when you want to do a more specific kind of analysis (perhaps you want to count bytes downloaded, in which case you only care about a couple event types).

FWIW, I think the main thing that would make either logger nicer and more stable would be to add typing/structs around the internal "fields" associated with each logged `Activity`. Currently every `Activity` comes with a `Logger::Fields`, which is an `std::vector` of things that are either numbers of strings. The specific meaning of the items in this vector depends on the `Activity`. IMO adding more structure to the fields internally and then reflecting it in the log output would make the output easier to consume. I think that's a bigger task though and hopefully shouldn't hold up merging something like a diff logger :)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-02-17-nix-team-meeting-minutes-33/25624/1

edolstra commented on src/libutil/logging-diffs.cc:
```diff
@@ -0,0 +1,228 @@
+struct DiffLogger : Logger {
+    Logger & prevLogger;
+
+    NixBuildState state;
```
This should probably be `Sync<NixBuildState> state` and then you can get rid of `std::mutex lock` and all that.

thomasjm replied:
Done

edolstra commented on src/libutil/logging-diffs.cc:
```diff
@@ -0,0 +1,228 @@
+            abort();
+}
+
+void to_json(json & j, const NixMessage & m) {
```
```suggestion
void to_json(json & j, const NixMessage & m)
 {
```

thomasjm replied:
Done

edolstra commented on src/libutil/logging-diffs.cc:
```diff
@@ -0,0 +1,228 @@
+    }
+}
+
+void add_pos_to_message(NixMessage & msg, std::shared_ptr<AbstractPos> pos)
```
```suggestion
static void addPosToMessage(NixMessage & msg, std::shared_ptr<AbstractPos> pos)
```

thomasjm replied:
Done

edolstra commented on src/libutil/logging-diffs.cc:
```diff
@@ -0,0 +1,228 @@
+    std::atomic_bool exited;
+    std::thread printerThread;
+
+    DiffLogger(Logger & prevLogger) : prevLogger(prevLogger),
+                                      last_sent(nullptr),
+                                      exitPeriodicAction(false),
+                                      exited(false),
+                                      printerThread(std::thread(&DiffLogger::periodicAction, this)) { }
```
```suggestion
    DiffLogger(Logger & prevLogger) 
        : prevLogger(prevLogger)
        , last_sent(nullptr)
        , exitPeriodicAction(false)
        , exited(false)
        , printerThread(std::thread(&DiffLogger::periodicAction, this)) 
        { }
```

thomasjm replied:
Done

edolstra commented on src/libutil/logging-diffs.hh:
```diff
@@ -0,0 +1,51 @@
+Logger * makeDiffLogger(Logger & prevLogger);
+
+struct ActivityState {
+    bool is_complete;
```
```suggestion
    bool isComplete = false;
```

fricklerhandwerk commented:
Discussed in Nix team meeting:

We're interested in the logger mechanism in general, but don't have the capacity to deal with the PR right now, and it's rather low priority. @roberth and @thufschmitt may look at it again in the future, but without commitment.

Profpatsch commented:
@thomasjm Maybe you can provide this `nix` patch in nixpkgs for interested users to try out on their systems? Then people could start using your tooling without blocking on upstream nix review.

thomasjm commented:
Thanks for considering this PR all!

@edolstra I addressed your review comments.

> Maybe you can provide this nix patch in nixpkgs for interested users to try out on their systems?

@Profpatsch hmm, that could be nice. Are you picturing an override on the `nix` derivation, or do you think we could provide it in such a way that it's cached on `cache.nixos.org`?

Profpatsch commented:
I’d just have a list of patches that can be added to nix via an option/package override and are kept up-to-date.
A subset could be added to all-packages.nix, but they should probably not be advertised too much outside of their respective PR/RFC, I guess.

thomasjm commented:
FWIW I've been keeping this rebased periodically. Just rebased on latest `master`.
